saleor/graphql/translations/mutations/utils.py

```diff
@@ -0,0 +1,467 @@
+    description = JSONString(description="Translated description." + RICH_CONTENT)
+
+
+class BaseBulkTranslateMutation(BaseMutation):
+    count = graphene.Int(
+        required=True, description="Returns how many translations were created/updated."
+    )
+
+    class Arguments:
+        error_policy = ErrorPolicyEnum(
+            required=False,
+            description="Policies of error handling. DEFAULT: "
+            + ErrorPolicyEnum.REJECT_EVERYTHING.name,
+        )
+
+    class Meta:
+        abstract = True
+
+    @classmethod
+    def __init_subclass_with_meta__(
+        cls,
+        base_model=None,
+        translation_model=None,
+        translation_fields=None,
+        base_model_relation_field=None,
+        result_type=None,
+        _meta=None,
+        **kwargs
+    ):
+        if not base_model:
+            raise ImproperlyConfigured(
+                "base model is required for bulk translation mutation"
+            )
+        if not translation_model:
+            raise ImproperlyConfigured(
+                "translation model is required for bulk translation mutation"
+            )
+        if not translation_fields:
+            raise ImproperlyConfigured(
+                "translation fields are required for bulk translation mutation"
+            )
+        if not result_type:
+            raise ImproperlyConfigured(
+                "result type is required for bulk translation mutation"
+            )
+        if not base_model_relation_field:
+            raise ImproperlyConfigured(
+                "base model relation field is required for bulk translation mutation"
+            )
+        if not _meta:
+            _meta = MutationOptions(cls)
+
+        _meta.base_model = base_model
+        _meta.translation_model = translation_model
+        _meta.translation_fields = translation_fields
+        _meta.base_model_relation_field = base_model_relation_field
+        _meta.result_type = result_type
+
+        doc_category_key = f"{base_model._meta.app_label}.{base_model.__name__}"
+        if "doc_category" not in kwargs and doc_category_key in DOC_CATEGORY_MAP:
+            kwargs["doc_category"] = DOC_CATEGORY_MAP[doc_category_key]
+
+        super().__init_subclass_with_meta__(_meta=_meta, **kwargs)
+
+    @classmethod
+    def clean_translations(cls, data_inputs, index_error_map):
+        """Retrieve the instance channel id for channel permission accessible check."""
+        raise NotImplementedError()
+
+    @classmethod
+    def pre_update_or_create(cls, translation, translation_fields):
+        return translation_fields
+
+    @classmethod
+    def get_base_objects(cls, cleaned_inputs_map: dict):
+        lookup = Q()
+        for data in cleaned_inputs_map.values():
+            if not data:
+                continue
+
+            if pk := data.get("id"):
+                lookup |= Q(pk=pk)
+            else:
+                lookup |= Q(external_reference=data.get("external_reference"))
```

Comment:
Are we checking if `external_reference` is not None anywhere? We have a check for non-nullability of `id` in `clean_translations`, but what would happen if we passed `null` as `externalReference` in the input?

Reply:
`validate_one_of_args_is_in_mutation` checks this so if `id` and `external_reference` will be `null` it will return a error